File: src/sensitivity_analysis/gen_figs.m
path = fullfile('sc_output', ...
    'sc_processed_tsim-300_tIinjstop-300_T-15_k-0-2');

ks = linspace(0.0, 2.0, 50);
sc_points = [3, 5, 7, 9];
var_percents = [5., 10., 20., 50.];
n_rows = numel(var_percents);
n_cols = numel(sc_points);

total_means = zeros(numel(ks), n_rows+n_cols);
total_vars = zeros(numel(ks), n_rows+n_cols);

% 4x4 plot, all combinations of CIs and SC points
figure('Renderer', 'painters', 'Position', [16 16 900 600])
plt_idx = 1;
for i = 1:n_rows
    ci = var_percents(i);
    for j = 1:n_cols
        sc = sc_points(j);
        mean_file = ['MEAN_CI', num2str(ci), '_SC_', num2str(sc), '.mat'];
        filepath = fullfile(path, mean_file);
        load(filepath);  % M variable with expected values
        total_means(:, plt_idx) = M;
        var_file = ['VAR_CI', num2str(ci), '_SC_', num2str(sc), '.mat'];
        filepath = fullfile(path, var_file);
        load(filepath);  % V variable with variances
        total_vars(:, plt_idx) = V;
        subplot(n_rows, n_cols, plt_idx);
        plot(ks, M, 'b.-');
        hold on
        plot(ks, M-sqrt(V), 'b-', ks, M+sqrt(V), 'b-');
        hold off
        xlabel('k');
        ylabel('E(<ISI>)(k)');
        title(['SC = ', num2str(sc), ', CI = ', num2str(ci), '%']);
        plt_idx = plt_idx + 1;
    end
end

% convergence of expected values and means of mean ISIs
figure('Renderer', 'painters', 'Position', [8 8 900 600])
start_plot_idx = 1;
end_plot_idx = start_plot_idx+3;
for i = 1:n_rows
    ci = var_percents(i);
    subplot(2, 2, i);
    plot(ks, total_means(:, start_plot_idx:end_plot_idx));
    xlabel('k');
    ylabel('E(<ISI>)(k)');
    title(['CI = ', num2str(ci), '%']);
    legend('3 SC points', '5 SC points', '7 SC points', '9 SC points');
    grid on;
    start_plot_idx = end_plot_idx+1;
    end_plot_idx = start_plot_idx+3;
end
figure('Renderer', 'painters', 'Position', [8 8 900 600])
start_plot_idx = 1;
end_plot_idx = start_plot_idx+3;
for i = 1:n_rows
    ci = var_percents(i);
    subplot(2, 2, i);
    plot(ks, total_vars(:, start_plot_idx:end_plot_idx));
    xlabel('k');
    ylabel('var(<ISI>)(k)');
    title(['CI = ', num2str(ci), '%']);
    legend('3 SC points', '5 SC points', '7 SC points', '9 SC points');
    grid on;
    start_plot_idx = end_plot_idx+1;
    end_plot_idx = start_plot_idx+3;
end

% sensitivity analysis for 5 sc points
figure('Renderer', 'painters', 'Position', [8 8 900 600])
for i = 1:n_rows
    ci = var_percents(i);
    anova_file = ['SA_ANOVA_indices_CI', num2str(ci), '_SC_5.mat'];
    filepath = fullfile(path, anova_file);
    load(filepath);
    subplot(2, 2, i);
    plot(ks, SS(:, 7:9));
    xlabel('k');
    ylabel('total effect sensitivity index');
    title(['CI = ', num2str(ci), '%']);
    legend('gNa sens. idx', 'gK sens. idx', 'gL sens. idx');
    grid on;
end

File: src/sensitivity_analysis/gen_figs.py
from itertools import cycle
import os

import matplotlib.pyplot as plt
import numpy as np
from scipy.io import loadmat


DATA_PATH = os.path.join('sc_output',
    'sc_processed_tsim-300_tIinjstop-300_T-15_k-0-2')
FIG_PATH = os.path.join('figures',
    'sc_processed_tsim-300_tIinjstop-300_T-15_k-0-2')
K_LIST = np.linspace(0, 2, 50)
SCP_LIST = [3, 5, 7, 9]
VAR_LIST = [5, 10, 20, 50]


def plotting_config(nrows=1, ncols=1):
    plt.style.use('seaborn-paper')
    plt.rcParams.update({
        'text.usetex': False,
        'font.family': 'serif',
        'font.size': 10,
        'axes.labelsize': 10,
        'grid.linewidth': 0.7,
        'legend.fontsize': 8,
        'xtick.labelsize': 8,
        'ytick.labelsize': 8,
        'figure.figsize': (4.774*nrows, 2.950*ncols),
        'figure.max_open_warning': 120,
    })


def load_data(ci, sc):
    mean_mat = loadmat(os.path.join(DATA_PATH, f'MEAN_CI{ci}_SC_{sc}.mat'))
    var_mat = loadmat(os.path.join(DATA_PATH, f'VAR_CI{ci}_SC_{sc}.mat'))
    return mean_mat, var_mat


def multiview(savefig=False):
    nrows = len(VAR_LIST)
    ncols = len(SCP_LIST)
    plotting_config(nrows, ncols)
    fig, ax = plt.subplots(nrows, ncols, sharex=True, sharey=True,
        squeeze=True)
    for ci_idx, ci in enumerate(VAR_LIST):
        for sc_idx, sc in enumerate(SCP_LIST):
            data = load_data(ci, sc) 
            mean_mISI = data[0]['M'].ravel()
            var_mISI = data[1]['V'].ravel()
            lb = mean_mISI - np.sqrt(var_mISI)
            ub = mean_mISI + np.sqrt(var_mISI)
            ax[ci_idx, sc_idx].plot(K_LIST, mean_mISI, label='<$mISI$>')
            ax[ci_idx, sc_idx].fill_between(K_LIST, lb, ub, alpha=0.2,
                label=f'{ci}% CI')
            if ci_idx == nrows-1:
                ax[ci_idx, sc_idx].set_xlabel('$k$')
            if sc_idx == 0:
                ax[ci_idx, sc_idx].set_ylabel('$mISI(k)$ [ms]')
            ax[ci_idx, sc_idx].legend(loc='best')
    plt.tight_layout()
    plt.show()
    if savefig:
        fig.savefig(fname=os.path.join(FIG_PATH, 'multiview.png'),
                format='png', bbox_inches='tight')


def sc_convergence(savefig=False):
    nfigs = len(VAR_LIST)
    plotting_config(nfigs, 2)
    fig, ax = plt.subplots(2, nfigs, squeeze=True)
    for ci_idx, ci in enumerate(VAR_LIST):
        for sc in SCP_LIST:
            data = load_data(ci, sc) 
            mean_mISI = data[0]['M'].ravel()
            var_mISI = data[1]['V'].ravel()
            ax[0, ci_idx].plot(K_LIST, mean_mISI, label=f'<$mISI$>, {sc} SC pts')
            ax[1, ci_idx].plot(K_LIST, var_mISI, label=f'$Var(mISI)$, {sc} SC pts')
            ax[1, ci_idx].set_xlabel('$k$')
            if ci_idx == 0:
                ax[0, ci_idx].set_ylabel('$mISI(k)$ [ms]')
                ax[1, ci_idx].set_ylabel('$Var(mISI)(k)$ [ms]')
            ax[0, ci_idx].legend(loc='best')
            ax[1, ci_idx].legend(loc='best')           
    plt.tight_layout()
    plt.show()
    if savefig:
        fig.savefig(fname=os.path.join(FIG_PATH, 'sc_convergence.png'),
                format='png', bbox_inches='tight')


def anova(savefig=False):
    nfigs = len(VAR_LIST)
    plotting_config(nfigs, 3)
    fig, ax = plt.subplots(3, nfigs, squeeze=True)
    sc = 5
    for ci_idx, ci in enumerate(VAR_LIST):
        data = loadmat(os.path.join(DATA_PATH, 
            f'SA_ANOVA_indices_CI{ci}_SC_{sc}.mat'))['SS']
        ax[0, ci_idx].plot(K_LIST, data[:, :3])
        ax[0, ci_idx].legend(['$g_{Na}$', '$g_{K}$', '$g_{L}$'])
        ax[1, ci_idx].plot(K_LIST, data[:, 3:6])
        ax[1, ci_idx].legend(['$g_{Na}$ $g_{K}$', '$g_{Na}$ $g_{L}$',
            '$g_{K}$ $g_{L}$'])
        ax[2, ci_idx].plot(K_LIST, data[:, 6:])
        ax[2, ci_idx].set_xlabel('$k$')
        ax[2, ci_idx].legend(['$g_{Na}$', '$g_{K}$', '$g_{L}$'], loc='lower right')
        if ci_idx == 0:
            ax[0, ci_idx].set_ylabel('$1^{st}$ order sensitivity')
            ax[1, ci_idx].set_ylabel('$2^{nd}$ order sensitivity')
            ax[2, ci_idx].set_ylabel('total effect sensitivity')
    plt.show()
    if savefig:
        fig.savefig(fname=os.path.join(FIG_PATH, 'anova.png'),
                format='png', bbox_inches='tight')


if __name__ == "__main__":
    savefig = True
    multiview(savefig)
    sc_convergence(savefig)
    anova(savefig)
